src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs

```diff
@@ -187,7 +187,30 @@ private async Task DoReceive()
 
                     input.Advance(bytesReceived);
 
-                    var flushTask = input.FlushAsync();
+                    ValueTask<FlushResult> flushTask;
+
+                    if (_stream.ReadsCompleted)
+                    {
+                        // If the data returned from ReadAsync is the final chunk on the stream then
+                        // flush data and end pipe together with CompleteAsync.
+                        //
+                        // Getting data and complete together is important for HTTP/3 when parsing headers.
+                        // It is important that it knows that there is no body after the headers.
+                        var completeTask = input.CompleteAsync(ResolveCompleteReceiveException(error));
+                        if (completeTask.IsCompletedSuccessfully)
```

Comment:
Why bother, we're already in a state machine

Reply:
Because the `ValueTask` needs to get converted to a `ValueTask<FlushResult>`. This path avoids the extra state machine that is in the local method.

(the reason to convert to `ValueTask<FlushResult>` is so `CompleteAsync` and `FlushAsync` use the same code path)

Reply:
It makes more sense if you see more of the method:

https://github.com/dotnet/aspnetcore/blob/ef7c86a56b8678157a8a523260b661f86090313b/src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs#L190-L227

Reply:
I see.